bin/client/src/cached_eigenda_provider.rs

```diff
@@ -0,0 +1,136 @@
+
+use rust_kzg_bn254::kzg::KZG;
+use rust_kzg_bn254::blob::Blob;
+use num::BigUint;
+
+
+#[derive(Debug, Clone)]
+pub struct CachedOracleEigenDAProvider<T: CommsClient> {
+    /// The preimage oracle client.
+    oracle: OracleEigenDAProvider<T>,
+    /// kzg proof witness
+    witness: EigenDABlobWitness,
+}
+
+impl<T: CommsClient> CachedOracleEigenDAProvider<T> {
+    /// Constructs a new oracle-backed EigenDA provider.
+    pub fn new(oracle: OracleEigenDAProvider<T>, witness: EigenDABlobWitness) -> Self {
+        Self { oracle, witness }
+    }
+}
+
+#[async_trait]
+impl <T: CommsClient + Sync + Send> EigenDABlobProvider for CachedOracleEigenDAProvider<T> {
+    type Error = OracleProviderError;
+
+    async fn get_blob(&mut self, cert: &Bytes) -> Result<Bytes, Self::Error> {
+        match self.oracle.get_blob(cert).await {
+            Ok(b) => {
+                let item_slice = cert.as_ref();
+                let cert_blob_info = match BlobInfo::decode(&mut &item_slice[4..]) {
+                    Ok(c) => c,
+                    Err(_) => return Err(OracleProviderError::Preimage(
+                        PreimageOracleError::Other("does not contain header".into(),
+                    ))),
+                };
+
+                let output = self.compute_and_save_witness(&b)?;
+                // make sure locally computed proof equals to returned proof from the provider
+                if output[..32] != cert_blob_info.blob_header.commitment.x[..] ||
+                   output[32..64] != cert_blob_info.blob_header.commitment.y[..]{
+                    return Err(OracleProviderError::Preimage(
+                        PreimageOracleError::Other("proxy commitment is different from computed commitment proxy".into())));
+                };
+
+                let commitment = Bytes::copy_from_slice(
+                    &[cert_blob_info.blob_header.commitment.x, cert_blob_info.blob_header.commitment.y]
+                    .concat());
+
+                let kzg_proof = Bytes::copy_from_slice(&output[64..128]);
+
+                // push data into witness
+                self.witness.write(b.clone().into(), commitment, kzg_proof.into());
+
+                Ok(b.into())
+            },
+            Err(e) => Err(e),
+        }
+    }
+}
+
+
+// nitro code https://github.com/Layr-Labs/nitro/blob/14f09745b74321f91d1f702c3e7bb5eb7d0e49ce/arbitrator/prover/src/kzgbn254.rs#L30
+impl<T: CommsClient + Sync + Send> CachedOracleEigenDAProvider<T> {
+    fn compute_and_save_witness(&mut self, blob: &[u8]) -> Result<Vec<u8>, OracleProviderError> {
+        // TODO remove the need for G2 access
+        // Add command line to specify where are g1 and g2 path
+        // In the future, it might make sense to let the proxy to return such
+        // value, instead of local computation
+        let mut kzg = match KZG::setup(
+            "resources/g1.32mb.point",
+            "",
+            "resources/g2.point.powerOf2",
```

Comment:
does this work? I thought the client doesn't have access to the fs? What if its running inside a zkvm..?